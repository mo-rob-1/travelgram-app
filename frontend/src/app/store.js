import { configureStore } from "@reduxjs/toolkit";
import authReducer from "../features/auth/authSlice";
import imageReducer from "../features/images/imageSlice";

export const store = configureStore({
  reducer: {
    auth: authReducer,
    images: imageReducer,
  },
});
