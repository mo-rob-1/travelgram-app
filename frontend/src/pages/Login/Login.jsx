import { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";
import { toast } from "react-toastify";
import { login, reset } from "../../features/auth/authSlice";
import {
  Container,
  StyledLink,
  Input,
  Button,
  Heading,
  HeadingWrapper,
  EmailFieldWrapper,
  FormWrapper,
  LinkWrapper,
  Section,
} from "./Login.styled";

function Login() {
  const [formData, setFormData] = useState({
    email: "",
    password: "",
  });

  const { email, password } = formData;

  const navigate = useNavigate();
  const dispatch = useDispatch();

  const { user, isLoading, isError, isSuccess, message } = useSelector((state) => state.auth);

  useEffect(() => {
    if (isError) {
      toast.error(message);
    }

    // Redirect when user is logged in
    if (isSuccess || user) {
      navigate("/");
    }

    dispatch(reset());
  }, [isError, isSuccess, user, message, navigate, dispatch]);

  const onChange = (e) => {
    setFormData((prevState) => ({
      ...prevState,
      [e.target.name]: e.target.value,
    }));
  };

  const onSubmit = (e) => {
    e.preventDefault();

    const userData = {
      email,
      password,
    };

    dispatch(login(userData));
  };

  if (isLoading) {
    return (
      <Section>
        <h2 style={{ marginBottom: "1rem" }}>Loading...</h2>
        <p>If the page is still loading, refresh page and try again (Ensure email and password are correct)</p>
      </Section>
    );
  }

  return (
    <>
      <Container>
        <FormWrapper>
          <HeadingWrapper>
            <Heading>Login</Heading>
            <p>Login to access your account</p>
          </HeadingWrapper>
          <section>
            <form onSubmit={onSubmit}>
              <EmailFieldWrapper>
                <label htmlFor="email">Email:</label>
                <Input
                  type="email"
                  data-testid="email-input"
                  id="email"
                  name="email"
                  value={email}
                  onChange={onChange}
                  required
                />
              </EmailFieldWrapper>

              <div>
                <label htmlFor="password">Password:</label>
                <Input
                  type="password"
                  data-testid="password-input"
                  id="password"
                  name="password"
                  value={password}
                  onChange={onChange}
                  required
                />
              </div>

              <div>
                <Button type="submit">Login</Button>
              </div>

              <LinkWrapper>
                <StyledLink to="/register">Sign Up Instead</StyledLink>
              </LinkWrapper>
            </form>
          </section>
        </FormWrapper>
      </Container>
    </>
  );
}

export default Login;
