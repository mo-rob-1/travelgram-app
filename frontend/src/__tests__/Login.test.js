import { render, screen, waitForDomChange } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Provider } from "react-redux";
import { store } from "../app/store";
import { BrowserRouter as Router } from "react-router-dom";
import Login from "../pages/Login/Login";

describe("Login", () => {
  it("renders the login page", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Provider store={store} history={history}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders every Login text", () => {
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );
    expect(screen.getAllByText("Login")).toBeTruthy();
  });

  it("renders email input", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    const emailInput = getByTestId("email-input");
    expect(emailInput).toBeTruthy();
  });

  it("renders password input", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    const passwordInput = getByTestId("password-input");
    expect(passwordInput).toBeTruthy();
  });

  it("renders login button", () => {
    const { getByRole } = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    const { getComputedStyle } = window;
    window.getComputedStyle = (elt) => getComputedStyle(elt);
    getByRole("button", { name: /login/i });
  });
});
