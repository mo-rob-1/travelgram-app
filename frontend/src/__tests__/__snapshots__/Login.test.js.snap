// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders the login page 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ doKmgW"
  >
    <div
      class="sc-gKXOVf lglQeP"
    >
      <div
        class="sc-hKMtZM RmZUw"
      >
        <h1
          class="sc-eCYdqJ cPjzgk"
        >
          Login
        </h1>
        <p>
          Login to access your account
        </p>
      </div>
      <section>
        <form>
          <div
            class="sc-jSMfEi cbYRGd"
          >
            <label
              for="email"
            >
              Email:
            </label>
            <input
              class="sc-gsnTZi dYzNv"
              data-testid="email-input"
              id="email"
              name="email"
              required=""
              type="email"
              value=""
            />
          </div>
          <div>
            <label
              for="password"
            >
              Password:
            </label>
            <input
              class="sc-gsnTZi dYzNv"
              data-testid="password-input"
              id="password"
              name="password"
              required=""
              type="password"
              value=""
            />
          </div>
          <div>
            <button
              class="sc-dkzDqf idLZHp"
              type="submit"
            >
              Login
            </button>
          </div>
          <div
            class="sc-iBkjds jxABVU"
          >
            <a
              class="sc-ftvSup eRLnRy"
              href="/register"
            >
              Sign Up Instead
            </a>
          </div>
        </form>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
